packages/node/src/integrations/tracing/nest.ts

```diff
@@ -44,7 +54,150 @@ interface MinimalNestJsApp {
 
 const INTEGRATION_NAME = 'Nest';
 
-export const instrumentNest = generateInstrumentOnce(INTEGRATION_NAME, () => new NestInstrumentation());
+const supportedVersions = ['>=8.0.0 <11'];
+
+const sentryPatched = 'sentryPatched';
+
+/**
+ * Represents an injectable target class in NestJS.
+ */
+interface InjectableTarget {
+  name: string;
+  sentryPatched?: boolean;
+  prototype: {
+    use?: (req: unknown, res: unknown, next: (error?: Error | unknown) => void) => void;
+  };
+}
+
+/**
+ * Helper checking if a target is already patched.
+ */
+function isPatched(target: InjectableTarget): boolean {
+  if (target.sentryPatched) {
+    return true;
+  }
+
+  addNonEnumerableProperty(target, sentryPatched, true);
+  return false;
+}
+
+/**
+ * Custom instrumentation for nestjs.
+ *
+ * This hooks into the @Injectable decorator, which is applied on class middleware, interceptors and guards.
+ */
+export class SentryNestInstrumentation extends InstrumentationBase {
+  public static readonly COMPONENT = '@nestjs/common';
+  public static readonly COMMON_ATTRIBUTES = {
+    component: SentryNestInstrumentation.COMPONENT,
+  };
+
+  public constructor(config: InstrumentationConfig = {}) {
+    super('sentry-nestjs', '1.0.0', config); // TODO: instrumentationVersion
+  }
+
+  /**
+   * Initializes the instrumentation by defining the modules to be patched.
+   */
+  public init(): InstrumentationNodeModuleDefinition {
+    const moduleDef = new InstrumentationNodeModuleDefinition(SentryNestInstrumentation.COMPONENT, supportedVersions);
+
+    moduleDef.files.push(this._getInjectableFileInstrumentation(supportedVersions));
+    return moduleDef;
+  }
+
+  /**
+   * Wraps the @Injectable decorator.
+   */
+  private _getInjectableFileInstrumentation(versions: string[]): InstrumentationNodeModuleFile {
+    return new InstrumentationNodeModuleFile(
+      '@nestjs/common/decorators/core/injectable.decorator.js',
+      versions,
+      (moduleExports: { Injectable: InjectableTarget }) => {
+        if (isWrapped(moduleExports.Injectable)) {
+          this._unwrap(moduleExports, 'Injectable');
+        }
+        this._wrap(moduleExports, 'Injectable', this._createWrapInjectable());
+        return moduleExports;
+      },
+      (moduleExports: { Injectable: InjectableTarget }) => {
+        this._unwrap(moduleExports, 'Injectable');
+      },
+    );
+  }
+
+  /**
+   * Creates a wrapper function for the @Injectable decorator.
+   *
+   * Wraps the use method to instrument nest class middleware.
+   */
+  private _createWrapInjectable() {
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    return function wrapInjectable(original: any) {
+      return function wrappedInjectable(options?: unknown) {
+        return function (target: InjectableTarget) {
+          // patch middleware
+          if (typeof target.prototype.use === 'function') {
+            // patch only once
```

Comment:
should this not be taken care of by the `isWrapped` / `unwrap` steps?

Reply:
The idea is that isWrapped checks whether the file exporting Injectable has been patched and this check is for concrete classes. Might not be necessary, just to play it safe. Will add a comment

Reply:
When would a specific class be wrapped more than once - can this happen?

Reply:
I am not sure I don't have a specific case in mind. Luca asked to add an extra guard here just to prevent running into funky edge cases